File: evals/retrievers/fasttext.ts
import type { Chunk } from '../lib/corpus.ts'
import { cosine } from '../lib/metrics.ts'
import { tokenize } from '../lib/tokenize.ts'
import type { Retriever } from './keyword.ts'

/**
 * Static word vectors averaged into a document vector — the pre-transformer way to do semantic
 * retrieval, and the point of comparison for what a contextual multilingual model actually buys.
 *
 * Weighted by IDF rather than plain mean: an unweighted average is dominated by the words every
 * chunk contains, so two documents about different subjects end up near each other because both
 * are made of Czech function words. IDF gives the weight to the words that distinguish them.
 *
 * Out-of-vocabulary words are skipped rather than zero-filled. A zero vector is not "no
 * information" — averaged in, it pulls every document toward the origin by an amount that depends
 * on how many words happen to be missing, which is a length artefact rather than a meaning.
 */
export function fasttextRetriever(
  chunks: Chunk[],
  vectors: Record<string, number[]>,
  dim: number,
  name = 'fasttext',
): Retriever {
  const df = new Map<string, number>()
  const docTokens = chunks.map((c) => tokenize(c.text))
  for (const terms of docTokens) {
    for (const t of new Set(terms)) {
      df.set(t, (df.get(t) ?? 0) + 1)
    }
  }
  const N = chunks.length
  const idf = (t: string) => Math.log(1 + (N - (df.get(t) ?? 0) + 0.5) / ((df.get(t) ?? 0) + 0.5))

  function embed(terms: string[]): Float32Array | null {
    const out = new Float32Array(dim)
    let total = 0
    for (const t of terms) {
      const v = vectors[t]
      if (!v) {
        continue
      }
      const w = idf(t)
      for (let i = 0; i < dim; i++) {
        out[i] += v[i] * w
      }
      total += w
    }
    // Every word was out of vocabulary: there is nothing to compare, and returning a zero vector
    // would score identically against everything rather than saying so.
    if (total === 0) {
      return null
    }
    for (let i = 0; i < dim; i++) {
      out[i] /= total
    }
    return out
  }

  const docVectors = docTokens.map(embed)

  return {
    name,
    search(query, k) {
      const q = embed(tokenize(query))
      if (!q) {
        return []
      }
      return docVectors
        .map((v, i) => ({ id: chunks[i].id, score: v ? cosine(q, v) : -1 }))
        .filter((s) => s.score > 0)
        .sort((a, b) => b.score - a.score)
        .slice(0, k)
        .map((s) => s.id)
    },
  }
}

File: evals/retrievers/fasttext.py
import math
from collections import Counter
from collections.abc import Mapping, Sequence

import numpy as np
from numpy.typing import NDArray

from lib.corpus import Chunk
from lib.tokenize import tokenize

Vector = NDArray[np.float32]


class FasttextRetriever:
    def __init__(
        self, chunks: list[Chunk], vectors: Mapping[str, Sequence[float]], dim: int, name: str = 'fasttext'
    ) -> None:
        self.name = name
        self.ids = [c.id for c in chunks]
        self.vectors = vectors
        self.dim = dim
        doc_tokens = [tokenize(c.text) for c in chunks]
        self.df = Counter(term for terms in doc_tokens for term in set(terms))
        self.n = len(chunks)
        doc_vectors = [self.embed(terms) for terms in doc_tokens]
        self.known = np.array([v is not None for v in doc_vectors])
        self.matrix = np.array([v if v is not None else np.zeros(dim, dtype=np.float32) for v in doc_vectors])
        self.matrix = self.matrix.astype(np.float64)
        self.norms = np.linalg.norm(self.matrix, axis=1)

    def idf(self, term: str) -> float:
        n = self.df[term]
        return math.log(1 + (self.n - n + 0.5) / (n + 0.5))

    def embed(self, terms: list[str]) -> Vector | None:
        out = np.zeros(self.dim, dtype=np.float32)
        total = 0.0
        for term in terms:
            v = self.vectors.get(term)
            if v is None:
                continue
            w = self.idf(term)
            out += np.asarray(v, dtype=np.float64) * w
            total += w
        if total == 0:
            return None
        out /= total
        return out

    def search(self, query: str, k: int) -> list[str]:
        q = self.embed(tokenize(query))
        if q is None:
            return []
        q64 = q.astype(np.float64)
        denominator = self.norms * float(np.linalg.norm(q64))
        with np.errstate(divide='ignore', invalid='ignore'):
            scores = np.where(self.known & (denominator > 0), (self.matrix @ q64) / denominator, -1.0)
        order: list[int] = np.argsort(-scores, kind='stable').tolist()
        return [self.ids[i] for i in order if scores[i] > 0][:k]
